core/unix/signal.c

```diff
@@ -1246,7 +1246,7 @@ signal_thread_inherit(dcontext_t *dcontext, void *clone_record)
          * FIXME: are current pending or blocked inherited?
          */
 #ifdef MACOS
-        if (record->app_thread_xsp != 0) {
+        if (record->app_thread_xsp == 0) {
```

Comment:
I don't really understand this code but you'll get heap oob/uaf asserts if this condition is `!=`.

Reply:
This is freeing the clone record.
See the comment in clone_record_t:
```
#ifdef MACOS
    /* XXX i#1403: once we have lower-level, earlier thread interception we can
     * likely switch to something closer to what we do on Linux.
     * This is used for bsdthread_create, where app_thread_xsp is NULL;
     * for vfork, app_thread_xsp is non-NULL and this is unused.
     */
```

This change != to == looks suspect.  Please double check vs this pasted comment: if bsdthread_create ends up different on aarch64 please update that comment; does it need separate handling a64 vs x86.

Reply:
Okay so please correct me if I'm wrong (and it's been like three months since I looked at this code, so I'm probably wrong), but `create_clone_record` sets app_thread_xsp to NULL iff the clone record is heap-allocated. So I think the only case in which we want to free it is if it's NULL.

https://github.com/DynamoRIO/dynamorio/blob/0502fd31a802466630c4358219924bc8a95366a5/core/unix/signal.c#L796-L803